source/vstd/std_specs/hash.rs

```diff
@@ -494,6 +489,227 @@ pub fn ex_hash_map_clear<Key, Value, S>(m: &mut HashMap<Key, Value, S>)
+#[verifier::external_type_specification]
+#[verifier::external_body]
+#[verifier::reject_recursive_types(Key)]
+#[verifier::reject_recursive_types(S)]
+pub struct ExHashSet<Key, S>(HashSet<Key, S>);
+
+impl<Key, S> View for HashSet<Key, S> {
+    type V = Set<Key>;
+
+    #[verifier::external_body]
+    closed spec fn view(&self) -> Set<Key>;
+}
+
+pub open spec fn spec_hash_set_len<Key, S>(m: &HashSet<Key, S>) -> usize;
+
+pub broadcast proof fn axiom_spec_hash_set_len<Key, S>(m: &HashSet<Key, S>)
+    ensures
+        obeys_key_model::<Key>() && builds_valid_hashers::<S>() ==> #[trigger] spec_hash_set_len(m)
+            == m@.len(),
+{
+    admit();
+}
+
+#[verifier::external_fn_specification]
+#[verifier::when_used_as_spec(spec_hash_set_len)]
+pub fn ex_hash_set_len<Key, S>(m: &HashSet<Key, S>) -> (len: usize)
+    ensures
+        len == spec_hash_set_len(m),
+{
+    m.len()
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_set_new<Key>() -> (m: HashSet<Key, RandomState>)
+    ensures
+        m@ == Set::<Key>::empty(),
+{
+    HashSet::<Key>::new()
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_set_with_capacity<Key>(capacity: usize) -> (m: HashSet<Key, RandomState>)
+    ensures
+        m@ == Set::<Key>::empty(),
+{
+    HashSet::<Key>::with_capacity(capacity)
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_set_reserve<Key, S>(m: &mut HashSet<Key, S>, additional: usize) where
+    Key: Eq + Hash,
+    S: BuildHasher,
+
+    ensures
+        m@ == old(m)@,
+{
+    m.reserve(additional)
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_set_insert<Key, S>(m: &mut HashSet<Key, S>, k: Key) -> (result: bool) where
+    Key: Eq + Hash,
+    S: BuildHasher,
+
+    ensures
+        obeys_key_model::<Key>() && builds_valid_hashers::<S>() ==> {
+            &&& m@ == old(m)@.insert(k)
+            &&& result == !old(m)@.contains(k)
+        },
+{
+    m.insert(k)
+}
+
+// The specification for `contains` has a parameter `key: &Q`
+// where you'd expect to find `key: &Key`. This allows for the case
+// that `Key` can be borrowed as something other than `&Key`. For
+// instance, `Box<u32>` can be borrowed as `&u32` and `String` can be
+// borrowed as `&str`, so in those cases `Q` would be `u32` and `str`
+// respectively.
+// To deal with this, we have a specification function that opaquely
+// specifies what it means for a set to contain a borrowed key of type
+// `&Q`. And the postcondition of `contains` just says that its
+// result matches the output of that specification function. But this
+// isn't very helpful by itself, since there's no body to that
+// specification function. So we have special-case axioms that say
+// what this means in two important circumstances: (1) `Key = Q` and
+// (2) `Key = Box<Q>`.
+pub spec fn set_contains_borrowed_key<Key, Q: ?Sized>(m: Set<Key>, k: &Q) -> bool;
+
+pub broadcast proof fn axiom_sets_contains_deref_key<Q>(m: Set<Q>, k: &Q)
```

Comment:
ObNitpick: Now that it's `set_contains_borrowed_key`, the axiom should probably be named `axiom_set_contains...`.